kernel/libs/aster-bigtcp/src/socket_table.rs

```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This module defines the socket table, which manages all TCP and UDP sockets,
+//! for efficiently inserting, looking up, and removing sockets.
+
+use alloc::{boxed::Box, collections::btree_map::BTreeMap, sync::Arc, vec::Vec};
+use core::net::Ipv4Addr;
+
+use jhash::{jhash_1vals, jhash_3vals};
+use smoltcp::wire::{IpAddress, IpEndpoint, IpListenEndpoint};
+use static_assertions::const_assert;
+
+use crate::{
+    ext::Ext,
+    socket::{TcpConnectionBg, TcpListenerBg, UdpSocketBg},
+    wire::PortNum,
+};
+
+pub type SocketHash = u32;
+
+/// A unique key for identifying a `TcpListener`.
+///
+/// Note that two `TcpListener`s cannot listen on the same address
+/// even if both sockets set SO_REUSEADDR to true,
+/// so there cannot be multiple listeners with the same `ListenerKey`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
```

Comment:
Remove `PartialOrd`/`Ord` since they are useless and I don't think they make sense for the `hash` field?

Reply:
The traits are removed from `ListenerKey`, but are still kept on `ConnectionKey`.

This is because `TcpBacklog` maintains a connecting map, so `ConnectionKey` should implement `Ord` to be the key of BtreeMap

```rust
pub struct TcpBacklog<E: Ext> {
    socket: Box<RawTcpSocket>,
    max_conn: usize,
    connecting: BTreeMap<ConnectionKey, TcpConnection<E>>,
    connected: Vec<TcpConnection<E>>,
}
```